Fix compilation and doctests with flintlib 3.2

tornaria opened the pull request:
Tested with flintlib 3.2-rc1.

May not work with flintlib 3.1.

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.

tornaria committed:
Fix compilation with flintlib 3.2

tornaria committed:
Fix doctests with flintlib 3.2

user202729 commented on src/sage/rings/polynomial/polynomial_element.pyx:
```diff
@@ -8398,19 +8398,19 @@ cdef class Polynomial(CommutativePolynomial):
-            [[1.00000000000 +/- ...e-12] + [+/- ...e-11]*I,
-             [1.0000000000 +/- ...e-12] + [+/- ...e-12]*I]
+            [[1.000000000... +/- ...] + [+/- ...]*I,
+             [1.000000000... +/- ...] + [+/- ...]*I]
```
how about using `abs tol` like what you used above?

tornaria replied:
I tried but it refuses to work because of the `doctest:...` above.

user202729 commented on src/sage/symbolic/ginac/numeric.cpp:
```diff
@@ -51,6 +51,7 @@
 
 #define PY_SSIZE_T_CLEAN
 #include <Python.h>
+#include "gmp.h"
```
Feel like flint bug to me? Surely the header should be usable standalone?

tobiasdiez replied:
I think it's by design: With
https://github.com/flintlib/flint/commit/3604bb2e4b630bc5a8104d17451badf972810979 one needs to include the gmp header before using `fmpz_get_mpz`.

tornaria committed:
Fix doctests to work with flintlib 3.1 as well

tornaria committed:
Restore compilation wit flintlib 3.1

tornaria committed:
Fix CIF sorting by forcing key=repr.

It turns out sorting in `ComplexIntervalField` is broken. In this case,
there are two elements that do not compare.
```
sage: x = PolynomialRing(CBF, 'x').gen()
sage: _, a, b, _ = sorted((x^4 - 3).roots(ComplexIntervalField(100),
multiplicities=False))
sage: a == b
False
sage: a < b
False
sage: a > b
False
sage: sorted([a,b])
[0.?e-37 + 1.31607401295249246081921890180?*I,
 0.?e-37 - 1.31607401295249246081921890180?*I]
sage: sorted([b,a])
[0.?e-37 - 1.31607401295249246081921890180?*I,
 0.?e-37 + 1.31607401295249246081921890180?*I]
```
